include/fiction/utils/layout_utils.hpp

```diff
@@ -397,6 +397,84 @@ CoordinateType random_coordinate(CoordinateType coordinate1, CoordinateType coor
+ * @return A vector containing all cells within the specified area.
+ */
+std::vector<siqad::coord_t> all_sidbs_in_spanned_area(const siqad::coord_t& cell_nw,
+                                                      const siqad::coord_t& cell_se) noexcept
+{
+    std::vector<siqad::coord_t> all_cells{};
+    const auto                  c1_cube          = siqad::to_fiction_coord<cube::coord_t>(cell_nw);
+    const auto                  c2_cube          = siqad::to_fiction_coord<cube::coord_t>(cell_se);
+    const uint64_t              total_cell_count = static_cast<uint64_t>(std::abs(c1_cube.x - c2_cube.x) + 1) *
+                                      static_cast<uint64_t>(std::abs(c1_cube.y - c2_cube.y) + 1);
+    all_cells.reserve(total_cell_count);
+
+    auto current_cell = cell_nw;
+
+    // collect all cells in the area (spanned by the nw `north-west` and se `south-east` cell) going from top to down
+    // from left to right.
+    while (current_cell <= cell_se)
+    {
+        all_cells.push_back(current_cell);
+        if (current_cell.x < cell_se.x)
+        {
+            current_cell.x += 1;
+        }
+        else if ((current_cell.x == cell_se.x) && current_cell.z == 0)
+        {
+            current_cell.z += 1;
+            current_cell.x = cell_nw.x;
+        }
+        else
+        {
+            current_cell.x = cell_nw.x;
+            current_cell.y += 1;
+            current_cell.z = 0;
+        }
+    }
+
+    return all_cells;
+}
+/**
+ * Modify a SiDB cell-level layout by adding cells based on specified indices.
+ *
+ * This function takes a cell-level SiDB layout, a list of all available cells, and a list of cell indices
+ * to add to the layout. It iterates through the provided indices, assigns NORMAL cell type to non-empty cells,
+ * and returns the modified layout.
+ *
+ * @tparam Lyt The cell-level SiDB layout type.
+ * @param lyt The SiDB cell-level layout to be modified.
+ * @param all_cells A vector containing all available cells.
+ * @param cell_indices A vector of indices indicating which cells to add.
+ * @return The modified layout with added cells.
+ */
+template <typename Lyt>
+Lyt add_cells_to_layout_based_on_indices(Lyt& lyt, const std::vector<typename Lyt::cell>& all_cells,
+                                         const std::vector<uint64_t>& cell_indices) noexcept
+{
+    static_assert(is_cell_level_layout_v<Lyt>, "Lyt is not a cell-level layout");
+    static_assert(has_sidb_technology_v<Lyt>, "Lyt is not an SiDB layout");
+    static_assert(has_siqad_coord_v<Lyt>, "Lyt is not based on SiQAD coordinates");
+
+    for (const auto i : cell_indices)
+    {
+        assert(i < all_cells.size() && "cell indices are out-of-range");
+        if (lyt.get_cell_type(all_cells[i]) == sidb_technology::cell_type::EMPTY)
+        {
+            lyt.assign_cell_type(all_cells[i], sidb_technology::cell_type::NORMAL);
+        }
+    }
+    return lyt;
+}
```

Comment:
This seems to be a niche use case that I would personally put in the respective algorithm's `detail` namespace. If this is needed a lot elsewhere, too, something might be wrong with the underlying code design principles.

Reply:
I agree. I was also unsure where to put it.